### ISSUE TITLE
chezmoi users: Please help with testing chezmoi2

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#1039]
What's new and what's changing in chezmoi v2: https://github.com/twpayne/chezmoi/blob/master/docs/CHANGES.md

If you're not interested, click the "Unsubscribe" button in this GitHub issue.

chezmoi v2 is close to being released and adds a huge number of new features while being almost completely backwards-compatible with v1. Thanks to testing by many contributors in #987, the code in chezmoi's `master` branch can be considered "release candidate 1".

You can grab pre-built binaries of [the latest build of `master`](https://github.com/twpayne/chezmoi/actions/workflows/main.yml?query=branch%3Amaster) (click on the most recent successful build, and then scroll down to the bottom of the page to find the build artifacts: there are Linux, macOS, and Windows binaries here). Or, you can build chezmoi from source (note that you need Go 1.16 now).

Please report any problems you find by either replying here or opening a new GitHub issue.

cc people who have opened [issues that are fixed in v2](https://github.com/twpayne/chezmoi/milestone/1?closed=1) and/or [already helped with testing](https://github.com/twpayne/chezmoi/issues/987) or [have a dotfiles repo using chezmoi with five or more stars](https://github.com/topics/chezmoi?o=desc&s=stars) or have been active recently: @3v1n0 @Croissong @JustinGrote @Legion2 @MurzNN @TsekNet @alker0 @angristan @beaussan @benmezger @caedes @caius @contrun @felipecrs @fwSmit @goooseman @gotgenes @jackmac92 @kitos9112 @m-rey @markstos @oscarbenedito @ptxmac @relaxdiego @renard @renemarc @schuerg @sersorrel @szorfein @terminalmage @vogu66 @woneill @zb140 @zuntik

If you're not interested, click the "Unsubscribe" button in this GitHub issue.

**Comment by ptxmac** [comment]
Quick suggestion, since you're requiring go1.16 already - could you add `darwin arm64` builds too? (Apple Silicon)

**Comment by ptxmac** [comment]
The builds from the "main" workflow report their version as `"version": "v0.0.0-SNAPSHOT-7a26aca"` - it might be useful to have this be `v2.0.0-SNAPSHOT-7a26aca` in case someone wishes to check for any new behaviour in a template

**Comment by ptxmac** [comment]
Added a bug report for the diff command: #1040

**Comment by 3v1n0** [comment]
Could you please push this version to the snap's edge/beta/candidate (the one you prefer) channel (or 2.0 track)?

**Comment by twpayne** [comment]
Thank you very much for the testing - really appreciated!

> Quick suggestion, since you're requiring go1.16 already - could you add darwin arm64 builds too? (Apple Silicon)

Done in 7d65398063a94f9d3d04d30bb41c6698736c89cf.

> The builds from the "main" workflow report their version as `"version": "v0.0.0-SNAPSHOT-7a26aca"` - it might be useful to have this be `v2.0.0-SNAPSHOT-7a26aca` in case someone wishes to check for any new behaviour in a template

Should be done in 6a2c5dc054f27138119bcc3d761cd90beb17fbed and f226fa385967466fb8e551933d8e087c33d59e03.

> Could you please push this version to the snap's edge/beta/candidate (the one you prefer) channel (or 2.0 track)?

Yes, will do. This will require a bit of release gymnastics.

> The build failed on my machine:

These are gpg integration tests, and there's definitely a bug here as the same tests fail on GitHub Actions on macOS and Windows. Thank you for reporting this, I will investigate.

> And if I manually run it it fails because no encryption is found:

What happens if you add

    encryption = "gpg"

to your config file?

> It also fails when I apply:
> `The `.config/chezmoi/scripts/macos_emacs-config.sh` is an old script but still referenced in `~/.config/chezmoi/chezmoistate.boltdb`

Can you indicate where this error message is coming from? I don't think it's from chezmoi.

Thank you again for testing - fixing these issues will make chezmoi really solid.

**Comment by renard** [comment]
> > And if I manually run it it fails because no encryption is found:
> 
> What happens if you add
> 
> ```
> encryption = "gpg"
> ```
> 
> to your config file?

Works better. Now I have other issues:

```
chezmoi: template: dot_config/resticprofile/sshconfig.tmpl:5:28: executing "dot_config/resticprofile/sshconfig.tmpl" at <.chezmoi.homedir>: map has no entry for key "homedir"
```
This must be template related since line in error is:
```
    IdentityFile {{ .chezmoi.homedir }}/.config/resticprofile/restic
```
`chezmoi data` reports that `homedir` is now spelt `homeDir`.

> 
> > It also fails when I apply:
> > `The `.config/chezmoi/scripts/macos_emacs-config.sh`is an old script but still referenced in`~/.config/chezmoi/chezmoistate.boltdb`
> 
> Can you indicate where this error message is coming from? I don't think it's from chezmoi.

Well in fact I am wrong with that. In fact any `run_*` scripts are failing. For some reason (I don't remember why, maybe because I wanted to have scripts managed by chezmoi as well) I have some scripts defined by:

```
mkdir -p "$(./chezmoi source-path)/dot_config/chezmoi/scripts/"
cat <<EOF > "$(./chezmoi source-path)/dot_config/chezmoi/scripts/run_foo"
#!/bin/sh

echo foo
EOF
chmod +x "$(./chezmoi source-path)/dot_config/chezmoi/scripts/run_foo"
```

This fails:
```chezmoi apply -v
panic: .config/chezmoi/scripts/foo: not in /Users/renard

goroutine 1 [running]:
github.com/twpayne/chezmoi/internal/chezmoi.AbsPath.MustTrimDirPrefix(...)
	/Users/renard/Src/chezmoi/internal/chezmoi/path.go:45
github.com/twpayne/chezmoi/internal/chezmoi.(*GitDiffSystem).trimPrefix(...)
	/Users/renard/Src/chezmoi/internal/chezmoi/gitdiffsystem.go:256
github.com/twpayne/chezmoi/internal/chezmoi.(*GitDiffSystem).RunScript(0xc0002e30e0, 0xc0006007e0, 0x1b, 0xc000672d50, 0x25, 0xc0000cd600, 0x14, 0x200, 0x50, 0x48)
	/Users/renard/Src/chezmoi/internal/chezmoi/gitdiffsystem.go:183 +0x33d
github.com/twpayne/chezmoi/internal/chezmoi.(*TargetStateScript).Apply(0xc000320560, 0x524c1d8, 0xc0002e30e0, 0x5246998, 0xc0007145e0, 0x523fba8, 0xc0004c17c0, 0x12, 0x4bd97e0, 0xc000534a50)
	/Users/renard/Src/chezmoi/internal/chezmoi/targetstateentry.go:195 +0x41d
github.com/twpayne/chezmoi/internal/chezmoi.(*SourceState).Apply(0xc000ad5ce0, 0x524c1d8, 0xc0002e30e0, 0x524c1d8, 0xc0002e30e0, 0x5246998, 0xc0007145e0, 0xc00003804d, 0xd, 0xc0006007e0, ...)
	/Users/renard/Src/chezmoi/internal/chezmoi/sourcestate.go:369 +0x3ec
github.com/twpayne/chezmoi/cmd.(*Config).applyArgs(0xc00031ce00, 0x524c1d8, 0xc0002e30e0, 0xc00003804d, 0xd, 0xc000075ca0, 0x0, 0x1, 0xc000540070, 0x120071a501, ...)
	/Users/renard/Src/chezmoi/cmd/config.go:418 +0x306
github.com/twpayne/chezmoi/cmd.(*Config).runApplyCmd(0xc00031ce00, 0xc0002b3680, 0xc000075ca0, 0x0, 0x1, 0x0, 0x0)
	/Users/renard/Src/chezmoi/cmd/applycmd.go:35 +0xcd
github.com/spf13/cobra.(*Command).execute(0xc0002b3680, 0xc000075c90, 0x1, 0x1, 0xc0002b3680, 0xc000075c90)
	/Users/renard/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:852 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002b2f00, 0xc0002b2f00, 0x0, 0x0)
	/Users/renard/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/renard/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
github.com/twpayne/chezmoi/cmd.(*Config).execute(0xc00031ce00, 0xc0000320d0, 0x2, 0x2, 0x0, 0x0)
	/Users/renard/Src/chezmoi/cmd/config.go:771 +0x7d
github.com/twpayne/chezmoi/cmd.runMain(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0000320d0, 0x2, ...)
	/Users/renard/Src/chezmoi/cmd/cmd.go:245 +0xe7
github.com/twpayne/chezmoi/cmd.Main(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0000320d0, 0x2, ...)
	/Users/renard/Src/chezmoi/cmd/cmd.go:85 +0x8e
main.main()
	/Users/renard/Src/chezmoi/main.go:17 +0x131```

**Comment by gotgenes** [comment]
>If possible it would be better to have backward compatibility and support both syntax with a deprecation warning. and remove this support in chezmoi 2.x or 3.

Alternatively, what if there is another 1.x release that announces these deprecation warnings? The [number of changes between 1 and 2](https://github.com/twpayne/chezmoi/blob/master/docs/CHANGES.md#changes-from-version-1) seems manageable.

**Comment by felipecrs** [comment]
Would be awesome if a few `alpha`, `beta`, or `rc` versions could be released to GitHub Releases as pre-releases so we could install them with:

```sh-session
sh -c "$(curl -fsSL https://git.io/chezmoi)" -- -t 2.0.0-alpha.1
```

**Comment by Legion2** [comment]
I get the following when running `chezmoi init`:
`chezmoi: template: chezmoi.yaml:1:21: executing "chezmoi.yaml" at <promptString "gpgRecipient">: error calling promptString: The handle is invalid.`

Was `promptString` removed? I didn't find anything in the changelog.

**Comment by twpayne** [comment]
> `error calling promptString: The handle is invalid.` Was `promptString` removed? I didn't find anything in the changelog.

No, `promptString` is still there, this is probably related to the TTY handling (chezmoi gets a TTY to read passwords). Does the error occur if you add `--no-tty` to the command? tl;dr this is probably a bug in TTY handling on Windows. Thanks for reporting it!

**Comment by Legion2** [comment]
Yes I'm using it on windows. When I add `--no-tty` there is no error and no prompt, but nothing happens.

**Comment by ptxmac** [comment]
Found another issue related to replacing files with symlinks: #1052

**Comment by twpayne** [comment]
@renard wrote:

> The build failed on my machine:

This should now be fixed with 5fd322cb6178ee2c0e6c7ab67fb6858adb247ab9.

Please do not ask how many WTFs were emitted during the fixing of this bug.

tl;dr it seems that gpg has a silent upper limit on the number of characters in a filename. The eight extra characters in the name of temporary directory used in the integration tests pushed gpg over that limit and gpg gave a cryptic error message. Removing eight characters from the temporary directory name made the test pass.

The gpg integration tests still don't pass on Windows. Currently, I suspect that this is because the integration tests are running on GitHub Actions that are using Windows underneath somehow, and this does not allow the creation of the UNIX Domain Socket that gpg uses to communicate with the agent that it spawns, but I'm probably on the wrong track.

**Comment by halostatue** [comment]
I’m not sure if this is just because I was trying this from a local build (`go run . diff`), but:

```
[ERROR] 2021/03/11 00:02:31 You are not currently signed in. Please run `op signin --help` for instructions
2021-03-11T00:02:31-05:00 ERR Output error="exit status 1" args=["op","get","document","aws-credentials: personal"] exitCode=1 output= path=/usr/local/bin/op systemTime=7.622 userTime=7.859
chezmoi: template: private_dot_aws/private_credentials.tmpl:2:3: executing "private_dot_aws/private_credentials.tmpl" at <onepasswordDocument (printf "aws-credentials: %v" .)>: error calling onepasswordDocument: op get document 'aws-credentials: personal': exit status 1
```

The first line is useful; the other lines are not. I’m presuming that this is because it’s a dev build (`chezmoi version dev`). If not, I’m not sure that the additional lines add any meaningful value (it made it harder for me to see the useful error).

**Comment by halostatue** [comment]
Running `chezmoi diff` (v1) is OK with my repo (currently private as I make sure I haven’t exposed any keys so I can safely rewrite the history until then), but (v2) results in a different error after fixing the fact that I wasn’t logged into `op`:

```
chezmoi: template: private_dot_config/gh/private_hosts.yml.tmpl:5:20: executing "private_dot_config/gh/private_hosts.yml.tmpl" at <keyring "github-gh-oauth" .github.user>: error calling keyring: assignment to entry in nil map
```

This is the contents of the specified template file:

```
{{- if (index . "github" "user") -}}
---
github.com:
    user: {{ .github.user }}
    oauth_token: {{ keyring "github-gh-oauth" .github.user }}
{{- end -}}
```

**Comment by twpayne** [comment]
Thanks for testing @halostatue!

> The first line is useful; the other lines are not.

Fair enough! https://github.com/twpayne/chezmoi/pull/1059/commits/cb4382005133624da045941196c62bf34fa7c52d compacts the first and third lines, and https://github.com/twpayne/chezmoi/pull/1059/commits/2ccc16a4170caff1ecab06422f67dde75165e23a removes the second line. The net result is that the error message would be:

    chezmoi: template: private_dot_aws/private_credentials.tmpl:2:3: executing
    "private_dot_aws/private_credentials.tmpl" at <onepasswordDocument (printf
    "aws-credentials: %v" .)>: error calling onepasswordDocument: op get document
    'aws-credentials: personal': exit status 1: [ERROR] 2021/03/11 00:02:31 You are
    not currently signed in. Please run `op signin --help` for instructions

This is perhaps still a bit verbose, but should include all the information for the user to debug the problem. The duplicated filename and template function name come from `text/template` and I don't know how to prevent the duplication. I really don't want to go down the path of parsing `op` error messages to decide what to include, as they are unlike to be stable over versions.

> error calling keyring: assignment to entry in nil map

I think this is due to Go's `text/template` eagerly evaluating things. I think there are two options here (warning: both untested):

1. Change your template to read:

    ```
    {{- if (index . "github" "user") -}}
    ---
    github.com:
        user: {{ index . "github" "user" }}
        oauth_token: {{ keyring "github-gh-oauth" (index . "github" "user") }}
    {{- end -}}
    ```

2. Change the template options so that missing keys return the zero value instead of raising an error, as described [in this section of the reference manual](https://github.com/twpayne/chezmoi/blob/82e7effdf2f737165f79180384f9a93afa60a4c3/docs/REFERENCE.md#template-execution).

That said, this behavior is not something that should have changed between v1 and v2, so I'm a bit surprised that the two versions give different results. Do you have an exact repro case?

**Comment by halostatue** [comment]
The error message is better, but I’m wondering if it might be possible to tell the difference between an error caused by an external tool like `op` and an error in the template itself. In the former case, you _might_ be able to construct a shorter but still meaningful error like:

```
chezmoi: template external error in private_dot_aws/private_credentials.tmpl
calling op get document 'aws-credentials: personal': [ERROR] 2021/03/11
00:02:31 You are not currently signed in. Please run `op signin --help` for
instructions
```

With respect to the `keyring` error: I get the same error when making either change. I am not getting any error when I call `go run . secret keyring get --service github-gh-oauth --user halostatue` (halostatue is the value of `.github.user`, obviously). Also, if I comment _out_ all of the keyring substitutions, I do not get this error.

**Comment by halostatue** [comment]
Further testing of the keyring error indicates that the issue is on line 31 of `cmd/keyringtemplatefuncs.go`: `c.keyring.cache[key] = password`.

The message is `assignment to entry in nil map` and it appears to be fixed with this patch:

```
diff --git c/cmd/keyringtemplatefuncs.go i/cmd/keyringtemplatefuncs.go
index e36c198..0d030a3 100644
--- c/cmd/keyringtemplatefuncs.go
+++ i/cmd/keyringtemplatefuncs.go
@@ -28,6 +28,11 @@ func (c *Config) keyringTemplateFunc(service, user string) string {
 		returnTemplateError(fmt.Errorf("%s %s: %w", service, user, err))
 		return ""
 	}
+
+	if c.keyring.cache == nil {
+		c.keyring.cache = make(map[keyringKey]string)
+	}
+
 	c.keyring.cache[key] = password
 	return password
 }
```

**Comment by twpayne** [comment]
> Further testing of the keyring error indicates that the issue is on line 31 of `cmd/keyringtemplatefuncs.go`: `c.keyring.cache[key] = password`.

Great catch! Yup, this is a definite bug (`c.keyring.cache` is never initialized in the current code in `master`). Would you like to submit a PR? Alternatively I will commit your fix with you credited as author. Thank you :)

**Comment by halostatue** [comment]
#1061 at your service.

**Comment by fwSmit** [comment]
I ran into this bug when running `chezmoi2 apply --source-path ...`. It gave an error that no tty was found, suggesting that some user input was needed. It turns out, some entirely different file was changed according to chezmoi2, even though I didn't change it.
```sh
❯ chezmoi2 apply
.XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? diff
diff --git a/.XCompose b/.XCompose
.XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? quit
❯ diff ./.XCompose ~/.local/share/chezmoi/dot_XCompose
❯ ls -l .XCompose
-rw-r--r-- 1 friso friso 3413 Sep 19 08:57 .XCompose
❯ ls -l ~/.local/share/chezmoi/dot_XCompose
-rw-r--r-- 1 friso friso 3413 Sep 19 08:57 ~/.local/share/chezmoi/dot_XCompose
```
I don't see any changes between the source file and the destination file, so maybe the internal cache is corrupted or something.

```
.XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? skip
.config/autostart has changed since chezmoi last wrote it [overwrite,all-overwite,skip,quit]? skip
...
.config/git/config has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? skip
.config/gtk-3.0/colors.css has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? quit
```
It seems like all files "have been changed since chezmoi last wrote it".
I consider this to be two bugs:
- Chezmoi incorrectly says the file has changed.
- `--source-path` should ignore all other files

**Comment by twpayne** [comment]
Thanks for reporting this @fwSmit.

> * `--source-path` should ignore all other files

I think this is already the case, I tried to reproduce the problem with ebdd6f4eb239c5b99fc22a8f9720ae8635ed9a62, but was unable to do so.

> Chezmoi incorrectly says the file has changed.

Given that chezmoi reports that all files have been modified, when I suspect that what's happened here is that chezmoi's state has been corrupted somehow. Would you mind running the following commands?

```
sha256sum ~/.XCompose
sha256sum ~/.local/share/chezmoi/dot_XCompose
chezmoi state dump | grep -A4 XCompose
stat ~/.XCompose
```

This should help reveal where the problem is.

**Comment by halostatue** [comment]
I can open this as a separate bug, but I’m trying to add a `modify_private_dot_netrc.tmpl` and all I’m getting is "chezmoi: exit status 1". How would I debug this?

Here’s the script that I’m trying:

```ruby
#!/usr/bin/env ruby

def add_if_missing(input, pattern, line)
  unless input.find { |r| r =~ pattern }
    input << line << "\n"
  end
end

input = STDIN.readlines

{{ if index . "github" "user" -}}
add_if_missing(
  input,
  /machine api.github.com login {{ .github.user }}/,
  "machine api.github.com login {{ .github.user }} {{ keyring "vim-rhubarb" .github.user }}"
)
{{- end }}

puts input.join
```

**Comment by fwSmit** [comment]
> I think this is already the case, I tried to reproduce the problem with ebdd6f4, but was unable to do so.

My bad, I was still on an older version and didn't test all bugs on the up-to-date version

```sh
> sha256sum ~/.XCompose
sha256sum ~/.local/share/chezmoi/dot_XCompose
ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80  /home/friso/.XCompose
ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80  /home/friso/.local/share/chezmoi/dot_XCompose

> chezmoi2 state dump | grep -A4 XCompose 
   "/home/friso/.XCompose": {
      "contentsSHA256": "ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80",
      "mode": 438,
      "type": "file"
    },

> stat ~/.XCompose
  File: /home/friso/.XCompose
  Size: 3413      	Blocks: 8          IO Block: 4096   regular file
Device: 10305h/66309d	Inode: 5768362     Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1000/   friso)   Gid: ( 1000/   friso)
Access: 2021-03-13 19:27:07.090972357 +0100
Modify: 2020-09-19 08:57:05.303266433 +0200
Change: 2020-09-19 08:57:05.309933080 +0200
 Birth: 2020-09-19 08:57:05.303266433 +0200
```

I don't see anything weird here, but maybe you do

**Comment by twpayne** [comment]
@halostatue Happy that you resolved the issue, but it sounds like chezmoi's error reporting could be improved here. Would you be able to share a case that reproduces the issue (e.g. using mixed quotes). I think that chezmoi should better forward errors from `modify_` scripts.

**Comment by twpayne** [comment]
> I don't see anything weird here, but maybe you do

@fwSmit it looks like the sha256 sums all line up. Is it possible that there's been a `umask` change? What does the output of `umask` print?

**Comment by fwSmit** [comment]
Umask just prints 022

**Comment by halostatue** [comment]
> @halostatue Happy that you resolved the issue, but it sounds like chezmoi's error reporting could be improved here. Would you be able to share a case that reproduces the issue (e.g. using mixed quotes). I think that chezmoi should better forward errors from `modify_` scripts.

Here’s two versions of the same script. First, the broken version:

```ruby
#!/usr/bin/env ruby

def add_if_missing(input, pattern, line)
  unless input.find { |r| r =~ pattern }
    input << line << "\n"
  end
end

input = STDIN.readlines

{{ if index . "github" "user" -}}
add_if_missing(
  input,
  /machine api.github.com login {{ .github.user }}/,
  "machine api.github.com login {{ .github.user }} {{ keyring "vim-rhubarb" .github.user }}"
)
{{- end }}

puts input.join
```

Now the fixed version:

```ruby
#!/usr/bin/env ruby

def add_if_missing(input, pattern, line)
  unless input.find { |r| r =~ pattern }
    input << line << "\n"
  end
end

input = STDIN.readlines

{{ if index . "github" "user" -}}
add_if_missing(
  input,
  /machine api.github.com login {{ .github.user }}/,
  'machine api.github.com login {{ .github.user }} {{ keyring "vim-rhubarb" .github.user }}'
)
{{- end }}

puts input.join
```

The only difference between the two is the quotes:

```diff
--- 1	2021-03-15 18:30:02.000000000 -0400
+++ 2	2021-03-15 18:29:54.000000000 -0400
@@ -12,7 +12,7 @@
 add_if_missing(
   input,
   /machine api.github.com login {{ .github.user }}/,
-  "machine api.github.com login {{ .github.user }} {{ keyring "vim-rhubarb" .github.user }}"
+  'machine api.github.com login {{ .github.user }} {{ keyring "vim-rhubarb" .github.user }}'
 )
 {{- end }}
 
```

**Comment by twpayne** [comment]
Thanks for the investigative work @halostatue :)

> How would I see what the _rendered_ script would be without running it? (That is, is there something like `chezmoi render modify_private_dot_netrc.tmpl` that I can run?)

You can use `chezmoi execute-template` for this, for example:

    chezmoi cd
    chezmoi execute-template < modify_private_dot_netrc.tmpl

These can be combined as:

    chezmoi execute-template < $(chezmoi source-path ~/.netrc)

I get the impression that `modify_` scripts are a powerful but hard-to-use tool, so I don't want to add specific tooling for them (like `chezmoi render`) until we develop some best practices for using them. In this case, the `modify_` script generates a Ruby script from a template, which is starting to get pretty meta. It might be easier to to manage all of `~/.netrc` with chezmoi, but depending on your exact requirements, this might not be an option for you.

**Comment by twpayne** [comment]
@fwSmit thanks for following up. I've opened #1066 to track the issue you reported.

**Comment by halostatue** [comment]
> Thanks for the investigative work @halostatue :)
> 
> > How would I see what the _rendered_ script would be without running it? (That is, is there something like `chezmoi render modify_private_dot_netrc.tmpl` that I can run?)
> 
> You can use `chezmoi execute-template` for this, for example:

> ```
> chezmoi execute-template < $(chezmoi source-path ~/.netrc)
> ```

That’s _perfect_ and exactly what I needed. What’s weird is that there may have been something else that I fixed at the same time as changing the quote types because I can’t reproduce the issue anymore. 

> I get the impression that `modify_` scripts are a powerful but hard-to-use tool, so I don't want to add specific tooling for them (like `chezmoi render`) until we develop some best practices for using them. In this case, the `modify_` script generates a Ruby script from a template, which is starting to get pretty meta. It might be easier to to manage all of `~/.netrc` with chezmoi, but depending on your exact requirements, this might not be an option for you.

No special tooling required; `execute-template` is exactly what I needed to see what I was wrong. The only way to really _modify_ a file in place with data from the environment (keyring, password vault, etc.) is to have a `.tmpl` of the execution script. The only thing I’d suggest at this point would be to point to `execute-template` for debugging templated scripts of _any_ type, not just _modify_ scripts.

**Comment by twpayne** [comment]
Thank you all @ptxmac @3v1n0 @renard @gotgenes @felipecrs @Legion2 @halostatue @fwSmit for all your testing here. Equally thank you to @zb140 @fwSmit @JustinGrote @MunifTanjim @caius in #987.

Your contributions have been incredibly helpful in catching a whole load of problems with chezmoi v2, and I particularly appreciate the time and effort that you have put in. Identifying and fixing these problems early, as you have done, will save the thousands of chezmoi users - and me - a huge amount of time in the future by hopefully providing software that "just works".

OK, time to release v2...

**Comment by twpayne** [comment]
OK, after all that fanfare, the release failed because the Snap Store is encountering problems:

```
   ⨯ release failed after 825.00s error=snapcraft packages: failed to publish artifacts: failed to push dist/chezmoi_2.0.0_linux_amd64.snap package: DEPRECATED: Use 'push' instead of 'upload'
Pushing chezmoi_2.0.0_linux_amd64.snap
After pushing, an attempt will be made to release to 'stable'
Preparing to push '/home/runner/work/chezmoi/chezmoi/dist/chezmoi_2.0.0_linux_amd64.snap' to the store.
The Snap Store encountered an error while processing your request: service unavailable (code 503).
The operational status of the Snap Store can be checked at https://status.snapcraft.io/

Error: The process '/opt/hostedtoolcache/goreleaser-action/0.159.0/x64/goreleaser' failed with exit code 1
```

https://status.snapcraft.io/

```
Mar 16, 2021
21:29
Snap Store degraded performance
The snap store is experiencing problems, we're investigating and will update once we know more.

Snap search is disabled for the time being, apologies for the inconvenience.

- Snap Store Team
```

This is quite funny. Will re-tag once the Snap Store is up again.
